### DISCUSSION TITLE
coupledValue vs coupledScalarValue

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#23971] [Q&A General] [answered]
Hello, I’m getting the following error when I try to add a coupled variable. 
```
/home/batodon/MOOSE/projects/ergs/src/userobjects/PorousFlowBrineSaltCO2.C: In constructor 'PorousFlowBrineSaltCO2::PorousFlowBrineSaltCO2(const InputParameters&)':
/home/batodon/MOOSE/projects/ergs/src/userobjects/PorousFlowBrineSaltCO2.C:30:16: error: 'coupledValue' was not declared in this scope; did you mean 'coupledScalar'?
   30 |     _solid_sat(coupledValue("solid_sat"))
```
I’ve correctly declared and retrieved the variable as follows:
```
  const VariableValue & _solid_sat;
```

```
params.addRequiredCoupledVar("solid_sat", "ss.");

_solid_sat(coupledValue("solid_sat"))
```

The error disappears when I replace `coupledValue` with `coupledScalarValue`, but I want to use the former. Could you please help? Thanks.

**Comment by loganharbour** [comment, marked as answer]
What UserObject type is `PorousFlowBrineSaltCO2`?

Many UOs cannot obtain variable values because they don't have the context of a "current" element/node/etc.

**Reply by batodon** [reply]
Hello @loganharbour, it is a `fluid_state` UO. I’m guessing it’s part of those that do not accept coupled variables. Correct?

**Reply by GiudGiud** [reply]
It's a fluid propery. 

It does not accept coupled variables, variables should be inputs to the routines, like density_from_p_T has pressure and temperature as inputs

coupled variables should not be attributes of this user object

**Reply by batodon** [reply]
oh well! Thanks.

**Reply by cpgr** [reply]
Isn't the solid saturation calculated by your user object given the pressure, temperature and salt mass fraction? I'm not sure why you want to add it as a variable here? Is it for an initial condition?

**Reply by batodon** [reply]
Hello @cpgr. No worries, I was testing a few ideas I had about `Xnacl` being equal to the solid saturation instead of `XEQ` in my `MassFraction`. I wanted the computed solid saturation at the previous time step to equal `Xnacl ` in the new time step. It’s a bad idea. Dw about it.
